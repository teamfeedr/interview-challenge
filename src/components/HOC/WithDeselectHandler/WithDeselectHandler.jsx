import React from 'react';
import PropTypes from 'prop-types';

export const getDeselectHandler = ({ state, updateState }) => ({ item }) => {
    const newItemsState = state.filter((stateItem) => stateItem.id !== item.id);

    updateState([...newItemsState]);
}

const WithDeselectHandler = ({ selectedItemsState, updateSelectedItemsState, WrappedComponent, ...childProps }) => {
    const handleDeselectItem = getDeselectHandler({ state: selectedItemsState, updateState: updateSelectedItemsState });

    return (<WrappedComponent handleDeselectItem={handleDeselectItem} {...childProps} />);
}

WithDeselectHandler.propTypes = {
    selectedItemsState: PropTypes.array.isRequired,
    updateSelectedItemsState: PropTypes.func.isRequired,
    WrappedComponent: PropTypes.func.isRequired
}

export default WithDeselectHandler;