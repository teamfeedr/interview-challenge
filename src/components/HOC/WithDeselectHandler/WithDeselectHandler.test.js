import React from "react";
import { shallow } from "enzyme";

import WithDeselectHandler, { getDeselectHandler } from "./WithDeselectHandler";

describe("the WithDeselectHandler HOC", () => {
  it("renders a wrapped component with the handleDeselectItem function prop", () => {
    const TestComponent = () => {
      <div />;
    };

    const rendered = shallow(
      <WithDeselectHandler
        selectedItemsState={[]}
        updateSelectedItemsState={() => {}}
        WrappedComponent={TestComponent}
      />
    );

    expect(rendered).toMatchSnapshot();
  });
});

describe("the getDeselectHandler function", () => {
  it("handles item deselection", () => {
    const state = [
      {
        id: 1
      },
      {
        id: 2
      }
    ];

    const updateState = jest.fn();

    const deselectItem = getDeselectHandler({ state, updateState });

    deselectItem({ item: { id: 1 } });

    expect(updateState).toHaveBeenCalledTimes(1);
    expect(updateState).toHaveBeenCalledWith([{ id: 2 }]);
  });
});
