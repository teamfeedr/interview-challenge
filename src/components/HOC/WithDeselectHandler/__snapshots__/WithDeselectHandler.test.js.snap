// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the WithDeselectHandler HOC renders a wrapped component with the handleDeselectItem function prop 1`] = `
<TestComponent
  handleDeselectItem={[Function]}
/>
`;
